server/src-lib/Hasura/GraphQL/Execute/Resolve.hs

```diff
@@ -0,0 +1,80 @@
+-- | Implements /variable resolution/ for GraphQL queries, which annotates the
+-- use site of each GraphQL variable with its value.
+module Hasura.GraphQL.Execute.Resolve
+  ( resolveVariables
+  ) where
+
+import           Hasura.Prelude
+
+import qualified Data.HashMap.Strict.Extended           as Map
+import qualified Data.HashSet                           as HS
+import qualified Data.List                              as L
+import qualified Data.Text                              as T
+import qualified Language.GraphQL.Draft.Syntax          as G
+
+import qualified Hasura.GraphQL.Transport.HTTP.Protocol as GH
+
+import           Hasura.GraphQL.Parser.Schema
+import           Hasura.RQL.Types.Error
+import           Hasura.SQL.Types
+
+resolveVariables
+  :: forall m fragments
+   . (MonadError QErr m, Traversable fragments)
+  => [G.VariableDefinition]
+  -> GH.VariableValues
+  -> G.SelectionSet fragments G.Name
+  -> m (G.SelectionSet fragments Variable)
+resolveVariables definitions jsonValues selSet = do
+  variablesByName <- Map.groupOnNE getName <$> traverse buildVariable definitions
+  uniqueVariables <- flip Map.traverseWithKey variablesByName
+    \variableName variableDefinitions ->
+      case variableDefinitions of
+        a :| [] -> return a
+        _      -> throw400 ParseFailed
+                   $ "multiple definitions for variable " <>> variableName
+  (selSet', usedVariables) <- flip runStateT mempty $
+    traverse (traverse (resolveVariable uniqueVariables)) selSet
+  let variablesByNameSet = HS.fromList . Map.keys $ variablesByName
+      jsonVariableNames = HS.fromList $ Map.keys jsonValues
+      isVariableValidationEnabled = False
```

Comment:
We should leave a comment / a TODO alongside those soon-to-be feature flags.

Reply:
I'll add the comment!

Reply:
Done